zstash/globus_utils.py

```diff
@@ -0,0 +1,273 @@
+            f"Token file {TOKEN_FILE} exists. We can try to load tokens from it."
+        )
+    else:
+        logger.info(
+            f"Token file {TOKEN_FILE} does NOT exist. This means we won't be able to load tokens from it."
+        )
+
+
+def get_local_endpoint_id(local_endpoint_id: Optional[str]) -> str:
+    ini = configparser.ConfigParser()
+    if ini.read(INI_PATH):
+        if "local" in ini.sections():
+            local_endpoint_id = ini["local"].get("globus_endpoint_uuid")
+            logger.info(
+                f"Setting local_endpoint_id based on {INI_PATH}: {local_endpoint_id}"
+            )
+    else:
+        ini["local"] = {"globus_endpoint_uuid": ""}
+        try:
+            with open(INI_PATH, "w") as f:
+                ini.write(f)
+            logger.info(f"Writing to empty {INI_PATH}")
+        except Exception as e:
+            logger.error(e)
+            sys.exit(1)
+    if not local_endpoint_id:
+        fqdn = socket.getfqdn()
+        if re.fullmatch(r"n.*\.local", fqdn) and os.getenv("HOSTNAME", "NA").startswith(
+            "compy"
+        ):
+            fqdn = "compy.pnl.gov"
+        for pattern in REGEX_ENDPOINT_MAP.keys():
+            if re.fullmatch(pattern, fqdn):
+                local_endpoint_id = REGEX_ENDPOINT_MAP.get(pattern)
+                logger.info(
+                    f"Setting local_endpoint_id based on FQDN {fqdn}: {local_endpoint_id}"
+                )
+                break
+    # FQDN is not set on Perlmutter at NERSC
+    if not local_endpoint_id:
+        nersc_hostname = os.environ.get("NERSC_HOST")
+        if nersc_hostname and (
+            nersc_hostname == "perlmutter" or nersc_hostname == "unknown"
+        ):
+            local_endpoint_id = REGEX_ENDPOINT_MAP.get(r"perlmutter.*\.nersc\.gov")
+            logger.info(
+                f"Setting local_endpoint_id based on NERSC_HOST {nersc_hostname}: {local_endpoint_id}"
+            )
+    if not local_endpoint_id:
+        logger.error(
+            f"{INI_PATH} does not have the local Globus endpoint set nor could one be found in REGEX_ENDPOINT_MAP."
+        )
+        sys.exit(1)
+    return local_endpoint_id
+
+
+def get_transfer_client_with_auth(
+    both_endpoints: List[Optional[str]],
+) -> TransferClient:
+    tokens = load_tokens()
+
+    # Check if we have stored refresh tokens
+    if "transfer.api.globus.org" in tokens:
+        token_data = tokens["transfer.api.globus.org"]
+        if "refresh_token" in token_data:
+            logger.info("Found stored refresh token - using it")
+            # Create a simple auth client for the RefreshTokenAuthorizer
+            auth_client = NativeAppAuthClient(ZSTASH_CLIENT_ID)
+            transfer_authorizer = RefreshTokenAuthorizer(
+                refresh_token=token_data["refresh_token"], auth_client=auth_client
+            )
+            transfer_client = TransferClient(authorizer=transfer_authorizer)
+            return transfer_client
```

Comment:
This resolves `(3) Start interactive zstash test transfer` & `(4) Start a second interactive zstash transfer`, and  from #339, as we no longer need to run a toy problem first. The only issue is that the very first run must be authenticated into. There's really no way of getting around that.